tsdb/head.go

```diff
@@ -1776,12 +1940,32 @@ func (s *memSeries) appendable(t int64, v float64) error {
 	return nil
 }
 
-func (s *memSeries) chunk(id int) *memChunk {
+// chunk returns the chunk for the id. If garbageCollect is true, it means that the returned *memChunk
+// (and not the chunkenc.Chunk inside it) can be garbage collected after it's usage.
+func (s *memSeries) chunk(id int, chunkDiskMapper *chunks.ChunkDiskMapper) (chunk *memChunk, garbageCollect bool) {
+	// ix represents the index of chunk in the s.mmappedChunks slice. The chunk id's are
+	// incremented by 1 when new chunk is created, hence (id - firstChunkID) gives the slice index.
+	// The max index for the s.mmappedChunks slice can be len(s.mmappedChunks)-1, hence if the ix
+	// is len(s.mmappedChunks), it represents the next chunk, which is the head chunk.
 	ix := id - s.firstChunkID
-	if ix < 0 || ix >= len(s.chunks) {
-		return nil
+	if ix < 0 || ix > len(s.mmappedChunks) {
+		return nil, false
+	}
+	if ix == len(s.mmappedChunks) {
+		return s.headChunk, false
 	}
-	return s.chunks[ix]
+	chk, err := chunkDiskMapper.Chunk(s.mmappedChunks[ix].ref)
+	if err != nil {
+		if err == chunks.ErrChunkDiskMapperClosed {
+			return nil, false
+		}
+		panic(err)
```

Comment:
do  we really want panic here?

Reply:
I dont see other clean ways to do it, returning error would cause a lot more places to be changed and something to be thought/designed over. Additionally, if there is an error, then there is something definitely wrong (data corruption), so a hard fail looks fine here.

Reply:
Not really sure, do we really want panic whole ingestion and query because of single map chunk file failed read?

Reply:
So the only blocker here is the [`Chunk` interface where we have `Iterator(Iterator) Iterator`](https://github.com/prometheus/prometheus/blob/532f7bbac92acedc1994b14130a3737cd88a7b6d/tsdb/chunkenc/chunk.go#L43-L68). We have this call here where we call that `.chunk` method https://github.com/prometheus/prometheus/blob/532f7bbac92acedc1994b14130a3737cd88a7b6d/tsdb/head.go#L1893-L1894
So in order to be able to propagate this error here, we need to change the interface and a whole lot of places. I will take a look after fixing the windows test fail.

Reply:
Turns out changing the interface is a big task :) do you have any suggestions here which would not involve changing the interface?

Reply:
metric, log and ignore? :thinking:

Reply:
Are we sure it's not recoverable error?

Reply:
Let's add TODO at least for now

Reply:
> Are we sure it's not recoverable error?

I think so. The error occurs whenever there is [corruption in data](https://github.com/prometheus/prometheus/blob/532f7bbac92acedc1994b14130a3737cd88a7b6d/tsdb/chunks/head_chunks.go#L434-L509). So in case of corruption, metric, log and ignore might not be the right thing to do, which will also lead to gaps in queries and maybe increase the time to recover the faulty disk.

I will add a TODO.